src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = {};
+  let stateAfterAction = {};
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        stateAfterAction = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...(stateHistory.at(-1) || { ...state }) };
+        stateAfterAction = Object.assign(currentState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        currentState = { ...(stateHistory.at(-1) || { ...state }) };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+
+        stateAfterAction = currentState;
+        break;
+    }
```

Comment:
Consider adding a default case to the switch statement to handle unexpected action types. This will ensure that any unexpected actions are handled gracefully.